src/mds/MDSDaemon.cc

```diff
@@ -821,6 +817,20 @@ void MDSDaemon::handle_mds_map(const cref_t<MMDSMap> &m)
         /* We are in the FSMap for the first time, and we're standby! */
         beacon.set_want_state(*mdsmap, new_state);
         dout(1) << "Monitors have assigned me to become a standby." << dendl;
+
+        /* The MDS has been added to the FSMap, now we can init the MgrClient */
+        mgrc.init();
+        auto r = mgrc.service_daemon_register("mds", name, {});
+        if (r < 0) {
+          derr << ": failed to register with manager for service status update" << dendl;
+          suicide();
+          return;
+        }
+
+        messenger->add_dispatcher_tail(&mgrc);
+        monc->sub_want("mgrmap", 0, 0);
+        monc->renew_subs(); /* MgrMap receipt drives connection to ceph-mgr */
+
```

Comment:
Can an MDS also start out in the creating state (instead of standby)?  (IIRC that used to be the case, once upon a time.)

Reply:
MDSMonitor calls `wait_for_finished_proposal` after adding the MDS as a standby. I'm not 100% sure it's not possible for an intervening MDSMonitor::tick to promote the new standby in the pending FSMap. I haven't seen it in my testing but I think it would involve a narrow race.

Perhaps it's just safer to change this so that any NULL -> X state initializes the mgrc.

Reply:
I've made this change.